data/src/main/java/com/teamagam/gimelgimel/data/notifications/PersistentConnectivityStatusRepositoryImpl.java

```diff
@@ -1,40 +1,82 @@
 package com.teamagam.gimelgimel.data.notifications;
 
-import com.teamagam.gimelgimel.data.base.ConsistentStatusEventRaiser;
 import com.teamagam.gimelgimel.data.base.repository.SubjectRepository;
 import com.teamagam.gimelgimel.domain.notifications.entity.ConnectivityStatus;
 import com.teamagam.gimelgimel.domain.notifications.repository.ConnectivityStatusRepository;
 
+import java.util.Timer;
+import java.util.TimerTask;
+
 import rx.Observable;
 
-/**
- * This repository updates the underlying state connectivity value only if the value is persistent
- * for the given timeframe.
- * Connectivity status is initialized to be CONNECTED
- */
 public class PersistentConnectivityStatusRepositoryImpl implements ConnectivityStatusRepository {
 
-    private SubjectRepository<ConnectivityStatus> mInnerRepo;
-    private ConsistentStatusEventRaiser<ConnectivityStatus> mConsistentEventRaiser;
-
+    private final SubjectRepository<ConnectivityStatus> mInnerRepo;
+    private Timer mTimer;
+    private long mTimeoutMillis;
+    private boolean mIsDisconnectTaskScheduled;
+    private ConnectivityStatus mCurrentStatus;
 
-    public PersistentConnectivityStatusRepositoryImpl(
-            ConnectivityStatus initialConnectivityStatus,
-            long consistentTimeFrameMS) {
+    public PersistentConnectivityStatusRepositoryImpl(ConnectivityStatus initStatus, long timeoutMillis) {
         mInnerRepo = SubjectRepository.createReplayCount(1);
-        mConsistentEventRaiser = new ConsistentStatusEventRaiser<>(
-                consistentTimeFrameMS,
-                initialConnectivityStatus,
-                consistentStatus -> mInnerRepo.add(consistentStatus));
+        mTimer = new Timer();
+        mTimeoutMillis = timeoutMillis;
+        mIsDisconnectTaskScheduled = false;
+        updateStatus(initStatus);
     }
 
     @Override
-    public void setStatus(ConnectivityStatus status) {
-        mConsistentEventRaiser.updateStatus(status);
+    public synchronized void setStatus(ConnectivityStatus newStatus) {
+        if (isConnectedUpdate(newStatus)) {
+            cancelDisconnectTask();
+            updateStatus(ConnectivityStatus.CONNECTED);
+        } else if (currentlyConnected() && noDisconnectTaskScheduledYet()) {
+            scheduleDisconnectTask();
+        }
     }
 
     @Override
     public Observable<ConnectivityStatus> getObservable() {
         return mInnerRepo.getObservable();
     }
+
+    private synchronized void updateStatus(ConnectivityStatus status) {
```

Comment:
There's no real need for this method to be synchronized as long as setStatus method is synchronized